src/Illuminate/Database/Eloquent/Builder.php

```diff
@@ -705,7 +705,13 @@ public function paginate($perPage = null, $columns = ['*'], $pageName = 'page',
 
         $perPage = $perPage ?: $this->model->getPerPage();
 
-        $results = ($total = $this->toBase()->getCountForPagination())
+        $query = $this->toBase();
+
+        $count = $query->distinct
+                        ? [ sprintf("%s.%s", $this->model->getTable(), $this->model->getKeyName()) ]
```

Comment:
I believe the table and column name need to be escaped/wrapped for compatibility with all database engines. I think that the wrapTable(...) and wrap(...) methods of the current grammar would be enough.